File: examples/python/ds/data/genre.py
from ds.base import *
from ds.model.genre import *
from ds.model._dataset import Dataset
from quartz.data.table import Table


class GenreData(Dataset.Genres):

    def __init__(self, db):
        self._table = GenreTable("genre", db)

    def add(self, x, debug:bool=False):
        if isinstance(x, Genre) and (x.id is None):
            x.id = self._table.insert(x, debug)
            return x
        elif self._replace(x, debug):
            return x
        self._table.insert(x, debug)
        return x

    def remove(self, x, debug:bool=False):
        return self._table.delete(x, debug)

    def push(self, x, field:Fields=None, debug:bool=False):
        if self._replace(x, True, debug):
            return x
        if isinstance(x, Genre) and isinstance(field, Fields):
            match field:
                case Fields.Name:
                    x = Table.Info("genre", [ ("id", x.id), ("name", x.name) ], 1)
                case _:
                    x = None
        return self._table.update(x, debug) if (x is not None) else None

    def pull(self, x:Genre, field:Fields=None, debug:bool=False):
        match field:
            case None:
                row = self._table.one(Table.Info("genre", [ ("id", x.id), ("name", None) ], 1), debug=debug)
                if (isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0):
                    x.name = row[0]
            case Fields.Name:
                row = self._table.one(Table.Info("genre", [ ("id", x.id), ("name", None) ], 1), debug=debug)
                if (isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0):
                    x.name = row[0]
        return x

    def has(self, x, debug:bool=False) -> bool:
        return self.ref(x, debug) is not None

    def clear(self, debug:bool=False):
        self._table.deleteAll(debug)

    def refb(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.begin(x, 'pk', order, limit, debug)

    def refn(self, cursor) -> Reference:
        return self.next(cursor, self._reference)

    def ref(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.one(x, self._reference, 'pk', order, limit, debug)

    def refs(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.all(x, self._reference, 'pk', order, limit, debug)

    def begin(self, x=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        return self._table.begin(x, columns, order, limit, debug)

    def next(self, cursor, builder=None):
        fn = self._create if (builder is None) else builder
        return self._table.next(cursor, fn)

    def one(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        fn = self._create if (builder is None) else builder
        return self._table.one(x, fn, columns, order, limit, debug)

    def all(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        fn = self._create if (builder is None) else builder
        return self._table.select(x, fn, columns, order, limit, debug)

    def _replace(self, x, allow_delete=False, debug:bool=False) -> bool:
        # "genre_book" book.id → genre.id
        if isinstance(x, BookGenreRef):
            if x.book_id and x.genre_id:
                return self._table.update(x, debug)
            elif allow_delete:
                return self._table.delete(x, debug)
            else:
                Error.invalid(f'"book_genre" replace', x)
        # "child_genre" genre.id → genre.id (recursive)
        if isinstance(x, SubgenreRef):
            if x.parent_id and x.child_id:
                return self._table.update(x, debug)
            elif allow_delete:
                return self._table.delete(x, debug)
            else:
                Error.invalid(f'"subgenre" replace', x)
        return False

    def _reference(self, x):
        return GenreRef(id=x[0])

    def _create(self, x):
        return Genre(id=x[0], name=x[1])


class GenreTable(Table):

    def _insert(self, x):
        # genre
        if isinstance(x, Genre):
            return Table.Info("genre", [ x.id, x.name ], 1, tag=("insert" if (x.id is None) else "insert_id"))

    def _update(self, x):
        # genre
        if isinstance(x, Genre):
            return Table.Info("genre", [ x.id, x.name ], 1, tag="update")
        # "genre_book" book.id → genre.id
        if isinstance(x, BookGenreRef):
            return Table.Info("book_genre", [ x.book_id, x.genre_id ], tag="replace")
        # "child_genre" genre.id → genre.id (recursive)
        if isinstance(x, SubgenreRef):
            return Table.Info("subgenre", [ x.parent_id, x.child_id ], tag="replace")

    def _delete(self, x):
        # "genre_book" book.id → genre.id
        if isinstance(x, BookGenreRef):
            return Table.Info("book_genre", [ ("book_id", x.book_id), ("genre_id", x.genre_id) ])
        # "child_genre" genre.id → genre.id (recursive)
        if isinstance(x, SubgenreRef):
            return Table.Info("subgenre", [ ("parent_id", x.parent_id), ("child_id", x.child_id) ])
        # pk
        if isinstance(x, int) or isinstance(x, GenreRef):
            return Table.Info("genre", [ int(x) ], tag="pk")
        # name
        if isinstance(x, str):
            return Table.Info("genre", [ x ], tag="name")

    def _select(self, x=None) -> tuple:
        # "genre_book" book.id → genre.id
        if isinstance(x, BookRef):
            return Table.Info("genre", [ int(x) ], tag="book")
        # "child_genre" genre.id → genre.id (recursive)
        if isinstance(x, ChildGenreRef):
            return Table.Info("genre", [ int(x) ], tag="child")
        # "parent_genre" genre.id → genre.id (recursive)
        if isinstance(x, ParentGenreRef):
            return Table.Info("genre", [ int(x) ], tag="parent")
        # pk
        if isinstance(x, int) or isinstance(x, GenreRef):
            return Table.Info("genre", [ int(x) ], tag="pk")
        # name (lookup)
        if isinstance(x, str):
            return Table.Info("genre", [ x ], tag="name")
        # all
        if x is None:
            return Table.Info("genre", tag="all")

File: examples/php/ds/data/genre.php
<?php
namespace books;
require_once __DIR__ . '/../base.php';
require_once __DIR__ . '/../model/genre.php';
require_once __DIR__ . '/../../common/data/base.php';
require_once __DIR__ . '/../../common/data/table.php';
require_once __DIR__ . '/../../common/error.php';

use quartz\Info, quartz\TagInfo, quartz\KeyInfo;
use quartz\Log;


class GenreData {
    private \quartz\Table $table;

    function __construct(\quartz\Database $db) {
        $this->table = new GenreTable("genre", $db);
    }

    function add($x, bool $debug=false) {
        if (($x instanceof model\Genre) && is_null($x->id)) {
            $x->id = $this->table->insert($x, $debug);
        }
        elseif (!$this->_replace($x, $debug)) {
            $this->table->insert($x, $debug);
        }
        return $x;
    }

    function push($x, bool $debug=false) {
        if (!$this->_replace($x, $debug)) {
            if (is_array($x) && (x[0] instanceof model\Genre) && (x[1] instanceof Fields)) {
                // Individual fields
                [$o, $f] = $x;
                if ($f == Fields::Name) {
                    $x = new KeyInfo("genre", [ ["id", $o->id] ], [ ["name", $o->name] ]);
                }
            }
            // Update
            return $this->table->update($x, $debug);
        }
    }

    function pull(model\Genre $x, ?Fields $field=null, bool $debug=false) {
        if (is_null($field)) {
            $row = $this->table->select(KeyInfo("genre", [ ["id", $x->id] ], [ ["name", $x->name] ]), one:true, debug:$debug);
            $x->name = $row[0];
        }
        elseif (field == Fields::Name) {
            $x->name = $this->table->select(KeyInfo("genre", [ ["id", $x->id] ], [ ["name", $x->name] ]), one:true, debug:$debug)[0];
        }
    }
    function remove($x, bool $debug=false) {
        return $this->table->delete($x, $debug);
    }

    function has($x, bool $debug=false): bool {
        return !is_null($this->ref($x, $debug));
    }

    function clear($debug=false) {
        $this->table->deleteAll(debug);
    }

    function refb($x=null, ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->begin($x, 'pk', $order, $limit, $debug);
    }

    function refn($cursor) {
        return $this->table->next($cursor, $this->reference(...));
    }

    function ref($x=null, ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->one($x, $this->reference(...), 'pk', $order, $limit, $debug);
    }

    function refs($x=null, ?array $order=null, ?string $limit=null, bool $debug=false): array {
        return $this->table->select($x, $this->reference(...), 'pk', $order, $limit, $debug);
    }

    function begin($x=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->begin($x, $columns, $order, $limit, $debug);
    }

    function next($cursor, ?callable $builder=null) {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->next($cursor, $fn);
    }

    function one($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->one($x, $fn, $columns, $order, $limit, $debug);
    }

    function all($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false): array {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->select($x, $fn, $columns, $order, $limit, $debug);
    }

    protected function _replace($x, bool $allow_delete=false, bool $debug=false) {
        // "genre_book" book.id → genre.id
        if ($x instanceof BookGenreRef) {
            if ($x->book_id && $x->genre_id) {
                return $this->table->update($x, $debug);
            }
            elseif ($allow_delete) {
                return $this->table->delete($x, $debug);
            }
            else {
                \quartz\Error::invalid('"book_genre" replace', $x);
            }
        }
        // "child_genre" genre.id → genre.id (recursive)
        if ($x instanceof SubgenreRef) {
            if ($x->parent_id && $x->child_id) {
                return $this->table->update($x, $debug);
            }
            elseif ($allow_delete) {
                return $this->table->delete($x, $debug);
            }
            else {
                \quartz\Error::invalid('"subgenre" replace', $x);
            }
        }
        return false;
    }

    protected function reference($x) {
        return new GenreRef(id:$x[0]);
    }

    protected function create($x) {
        return new model\Genre(id:$x[0], name:$x[1]);
    }
}


class GenreTable extends \quartz\Table {

    function _insert($x): ?Info {
        // genre
        if ($x instanceof model\Genre) {
            return new Info("genre", [ $x->id, $x->name ]);
        }
        return null;
    }

    function _update($x): ?Info {
        // genre
        if ($x instanceof model\Genre) {
            return new Info("genre", [ $x->name, $x->id ]);
        }
        // "genre_book" book.id → genre.id
        if ($x instanceof BookGenreRef) {
            return new Info("book_genre", [$x->book_id, $x->genre_id ], true);
        }
        // "child_genre" genre.id → genre.id (recursive)
        if ($x instanceof SubgenreRef) {
            return new Info("subgenre", [$x->parent_id, $x->child_id ], true);
        }
        return null;
    }

    function _delete($x): ?Info {
        // "genre_book" book.id → genre.id
        if ($x instanceof BookGenreRef) {
            return KeyInfo("book_genre", [ ["book_id", $x->book_id], ["genre_id", $x->genre_id] ]);
        }
        // "child_genre" genre.id → genre.id (recursive)
        if ($x instanceof SubgenreRef) {
            return KeyInfo("subgenre", [ ["parent_id", $x->parent_id], ["child_id", $x->child_id] ]);
        }
        // pk
        if (($x instanceof int) || ($x instanceof GenreRef)) {
            return new TagInfo("genre", "pk", [ $x->id ]);
        }
        // name
        if ($x instanceof str) {
            return new TagInfo("genre", "name", [ x ]);
        }
        return null;
    }

    function _select($x=null): ?Info {
        // "genre_book" book.id → genre.id
        if ($x instanceof BookRef) {
            return new TagInfo("genre", "book", [ $x->id ]);
        }
        // "child_genre" genre.id → genre.id (recursive)
        if ($x instanceof ChildGenreRef) {
            return new TagInfo("genre", "child", [ $x->id ]);
        }
        // "parent_genre" genre.id → genre.id (recursive)
        if ($x instanceof ParentGenreRef) {
            return new TagInfo("genre", "parent", [ $x->id ]);
        }
        // pk
        if (($x instanceof int) || ($x instanceof GenreRef)) {
            return new TagInfo("genre", "pk", [ $x->id ]);
        }
        // name (lookup)
        if ($x instanceof str) {
            return new TagInfo("genre", "name", [ $x ]);
        }
        // all
        if (is_null($x)) {
            return new TagInfo("genre", "all");
        }
        return null;
    }
}

?>